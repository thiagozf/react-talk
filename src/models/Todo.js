// @flow

// $FlowFixMe
import { Record, type RecordFactory } from 'immutable';

export type TTodo = {
  id: string,
  text: string,
  done: boolean,
};

const Todo: RecordFactory<TTodo> = Record({
  id: '',
  text: '',
  done: false,
});

export default Todo;
