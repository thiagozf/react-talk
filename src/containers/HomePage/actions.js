// @flow

import {
  LOAD_TODOS,
  LOAD_TODOS_SUCCESS,
  LOAD_TODOS_ERROR,
  TOGGLE_TODO,
} from 'containers/HomePage/constants';

import { type TTodo } from 'models/Todo';

export const loadTodos = () => ({ type: LOAD_TODOS });

export const todosLoaded = (todos: Array<TTodo>) => ({
  type: LOAD_TODOS_SUCCESS,
  todos,
});

export const todosLoadingError = (error: string) => ({
  type: LOAD_TODOS_ERROR,
  error,
});

export const toggleTodo = (id: string, done: boolean) => ({
  type: TOGGLE_TODO,
  id,
  done,
});
