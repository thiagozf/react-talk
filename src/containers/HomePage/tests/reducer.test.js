import { fromJS } from 'immutable';

import Todo from 'models/Todo';

import homeReducer from '../reducer';
import { loadTodos, todosLoadingError, todosLoaded, toggleTodo } from '../actions';

describe('homeReducer tests', () => {
  it('should load initial state', () => {
    const state = homeReducer(undefined, { type: 'OTHER' });
    expect(state).toMatchSnapshot();
  });

  it('should load todos', () => {
    const state = homeReducer(undefined, loadTodos());
    expect(state.get('loading')).toBeTruthy();
    expect(state.get('error')).toBeFalsy();
    expect(state.getIn(['data', 'todos'])).toBeFalsy();
  });

  it('should store errors', () => {
    const error = 'homeReducer#error';
    const state = homeReducer(undefined, todosLoadingError(error));
    expect(state.get('loading')).toBeFalsy();
    expect(state.get('error')).toBe(error);
  });

  it('should store successfully loaded todos', () => {
    const todos = [new Todo({ id: 'test', text: 'testing stuff', done: true })];
    const state = homeReducer(undefined, todosLoaded(todos));
    expect(state.get('loading')).toBeFalsy();
    expect(state.get('error')).toBeFalsy();
    expect(state.getIn(['data', 'todos'])).toBe(todos);
  });

  it('should toggle todo\'s "done" state', () => {
    const initialState = fromJS({
      loading: false,
      error: false,
      data: {
        todos: [
          new Todo({ id: '1', text: 'test1', done: false }),
          new Todo({ id: '2', text: 'test2', done: false }),
        ],
      },
    });
    const state = homeReducer(initialState, toggleTodo('2', true));
    expect(state.getIn(['data', 'todos', 0, 'done'])).toBeFalsy();
    expect(state.getIn(['data', 'todos', 1, 'done'])).toBeTruthy();
  });
});
