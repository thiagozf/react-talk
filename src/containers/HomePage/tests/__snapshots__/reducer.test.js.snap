// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`homeReducer tests should load initial state 1`] = `
Immutable.Map {
  loading: false,
  error: false,
  data: Immutable.Map {
    todos: false,
  },
}
`;
